api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/AbstractPool.java

```diff
@@ -52,17 +62,54 @@ public C getByName(String name) {
 
     public void printNames() {
         for (Map.Entry<String, C> entry : components.entrySet()) {
-            log.debug("C in pool: '" + entry.getKey() + "'");
+            log.debug(format("%s in pool: '%s'", getType().getName(), entry.getKey()));
+        }
+    }
+
+    public void add(String uri, C component) {
+        String name = component.getName();
+        log.info(format("Adding component %s with URI %s", name, uri));
+        if (isInModel(uri)) {
+            synchronized (mutex) {
+                C oldComponent = components.put(name, component);
+                if (oldComponent != null && oldComponent.hasClients()) {
```

Comment:
The same, I would add it to obsolete components and let it check if it has clients and run dereferencing